### PR TITLE
Add c8y-firmware-plugin

### PR DESCRIPTION
**Pull request by rina23q**
## Proposed changes
This PR should implement as described in the ticket #1696. 
The following memo might help you to understand the code better.

* `c8y-firmware-plugin --init` creates 3 required directories: `/var/tedge/cache`, `/var/tedge/firmware`, and `/var/tedge/file-transfer`. Make sure that `/var/tedge/file-transfer` can be also created by `tedge-agent` upon the first POST request.
* The downloaded file will be stored in `/tmp` first, then move to `/var/tedge/cache`. If `/var/tedge/cache` doesn't exist, ~~this plugin creates the directory in runtime~~ **returns error**.
* The file transfer dir `/var/tedge/file-transfer/<child-id>/firmware_update` will keep a symlink of the file. ~~If this directory doesn't exist, this plugin creates the directory in runtime.~~ If `/var/tedge/file-transfer` directory doesn't exist, returns error. `/<child-id>/firmware_update` directory will be created in runtime.
* The name of downloaded file and the symlink are sha256 of the file URL from the server (not localhost!).
* The `sha256` field of request and response JSON is the sha256 of file itself. Checksum if they downloaded correct file.
* The operation timeout is configurable by tedge CLI. e.g. ~~`tedge config set firmware.timeout 200`~~ `tedge config set firmware.child.update.timeout 200`.
* If `status` field in the response from child device is invalid but operation ID is correct, the operation status is treated `None` internally. From user, it will be an error.

### NOT in this PR's scope
* Creation of supported operation file named `c8y_Firmware`. This will be created by user.
* Heartbeat from child device to extend the operation timer.
* Case-insensitive status on child device response.

## TODO before merge
- [x] Add uuid and attempt to operation payload.
- [x] Add mechanism to read/write a status operation to a file to keep persistent server url - localhost url mapping.
- [ ] Update "after help text". The output of `c8y-firmware-plugin --help`.
- [x] Add tests!!!
- [x] `--init` creates directories.
- [x] Error handling. If an error occurs, make c8y_Firmware operation failed.
- [x] Rename the tedge config key from `firmware.timeout` to `firmware.child.update.timeout`.

## Types of changes
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
#1696

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments
Directory/file permissions under `/var/tedge` and its security.

* `/var/tedge` is created by `tedge-agent --init` owned by `tedge:tedge`.
* `/var/tedge/file-transfer` is created by `tedge-agent` owned by `tedge:tedge upon `PUT` request. (timing of child device bootstrapping of config plugin)
* `/var/tedge/file-transfer/<child-id>` is created by `tedge-agent` owned by `tedge:tedge` upon `PUT` request. (timing of child device bootstrapping of config plugin)
*  `/var/tedge/file-transfer/<child-id>/config_update` is created by `c8y-configuration-plugin` owned by `root:root`.

This plugin needs them. Now c8y-firmware-plugin is supposed to be run by `tedge` user, therefore, they will be owned by `tedge:tedge`. But it is a security risk as any one can modify via tedge-agent http server.
* `/var/tedge/cache/<file>`
* `/var/tedge/file-transfer/<child-id>/firmware_update/<symlink>`

### CODE REVIEW

File: plugins/c8y_firmware_plugin/src/firmware_manager.rs [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        timeout_sec: Duration,
+    ) -> Result<Self, anyhow::Error> {
+        let mqtt_client = Self::create_mqtt_client(mqtt_port).await?;
+
+        let c8y_request_topics = C8yTopic::SmartRestRequest.into();
+        let health_check_topics = health_check_topics(PLUGIN_SERVICE_NAME);
+        let firmware_update_response_topics =
+            TopicFilter::new_unchecked(FIRMWARE_UPDATE_RESPONSE_TOPICS);
+
+        let firmware_download_manager = FirmwareDownloadManager::new(
```
Comment (albinsuresh) [inline plugins/c8y_firmware_plugin/src/firmware_manager.rs:54]: Although I understand that this design was inspired by the `c8y-config-plugin`, a separate `FirmwareDownloadManager` abstraction is less useful/relevant here as this plugin doesn't have a "upload" notion which would required another `FirmwareUploadManager`. So, it'd be simpler to keep everything in the `FirmwareDownloadManager` in this `FirmwareManager` struct itself, to avoid an unnecessary level of indirection.

File: plugins/c8y_firmware_plugin/src/firmware_manager.rs [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    async fn process_mqtt_message(&mut self, message: Message) -> Result<(), anyhow::Error> {
+        if self.health_check_topics.accept(&message) {
+            send_health_status(&mut self.mqtt_client.published, PLUGIN_SERVICE_NAME).await;
+            return Ok(());
+        } else if self.firmware_update_response_topics.accept(&message) {
+            self.handle_child_device_firmware_operation_response(&message)
+                .await?
+        } else if self.c8y_request_topics.accept(&message) {
+            for smartrest_message in collect_smartrest_messages(message.payload_str()?) {
+                let result = match smartrest_message.split(',').next().unwrap_or_default() {
```
Comment (albinsuresh) [inline plugins/c8y_firmware_plugin/src/firmware_manager.rs:114]:
```suggestion
                let result = c8y_api::smartrest::message::get_smartrest_template_id(smartrest_message) {
```
An API exists for this already.

File: plugins/c8y_firmware_plugin/src/firmware_manager.rs [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+                    "515" => {
+                        match SmartRestFirmwareRequest::from_smartrest(smartrest_message.as_str()) {
+                            Ok(firmware_request) => {
+                                self.firmware_download_manager
+                                    .handle_firmware_download_request(firmware_request)
+                                    .await
+                            }
+                            Err(_) => {
+                                error!("Incorrect SmartREST payload: {smartrest_message}");
```
Comment (albinsuresh) [inline plugins/c8y_firmware_plugin/src/firmware_manager.rs:123]:
```suggestion
                                error!("Incorrect c8y_Firmware SmartREST payload: {smartrest_message}");
```

File: plugins/c8y_firmware_plugin/src/child_device.rs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    pub fn operation_request_topic(&self) -> Topic {
+        Topic::new_unchecked(&format!(
+            "tedge/{}/commands/req/firmware_update",
+            self.child_id
+        ))
+    }
+
+    pub fn operation_request_payload(&self, url: &str) -> Result<String, anyhow::Error> {
```
Comment (albinsuresh) [inline plugins/c8y_firmware_plugin/src/child_device.rs:89]:
```suggestion
    pub fn operation_request_payload(&self, url: &str) -> Result<String, FirmwareManagementError> {
```

File: plugins/c8y_firmware_plugin/src/common.rs [hunk 5]
```diff
@@ -0,0 +1,61 @@
+use crate::download::DownloadFirmwareStatusMessage;
```
Comment (albinsuresh) [inline plugins/c8y_firmware_plugin/src/common.rs:1]: I'd merge the content of this module and the `child_device.rs` module into a single `messages.rs` module as most of the stuff in these crates are about the message definitions. Having a separate `child_device.rs` module doesn't make much sense here either as this plugin is **only** about child devices.

File: plugins/c8y_firmware_plugin/src/download.rs [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+            smartrest_request.device,
+            smartrest_request.name,
+            smartrest_request.version,
+            smartrest_request.url,
+        );
+
+        if smartrest_request.device == self.tedge_device_id {
+            warn!("c8y-firmware-plugin does not support firmware operation for the main tedge device. \
+            Please define a custom operation handler for the c8y_Firmware operation.");
+            Ok(())
```
Comment (albinsuresh) [inline plugins/c8y_firmware_plugin/src/download.rs:96]: Seeing that `Ok(())` here surprised me a bit as I thought that it'd make the operation succeed in the cloud. But since that's not the case, it's fine. May be worth adding a comment that we're ignoring this request so that some other plugin on the machine can handle it.

File: plugins/c8y_firmware_plugin/src/download.rs [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+
+        // Add a pair of local url and external url to the hashmap.
+        self.url_map
+            .insert(file_transfer_url.clone(), firmware_url.to_string());
+
+        let file_sha256 = try_digest(transfer_dest.as_path())?;
+        let firmware_entry =
+            FirmwareEntry::new(firmware_name, firmware_version, firmware_url, &file_sha256);
+        let firmware_op_req = FirmwareOperationRequest::new(child_id, firmware_entry);
+        let firmware_update_req_msg = Message::new(
```
Comment (albinsuresh) [inline plugins/c8y_firmware_plugin/src/download.rs:185]: Minor: It would have been better if the `FirmwareOperationRequest` operation implemented `Into<Message>` to make this call seemless.

File: plugins/c8y_firmware_plugin/src/download.rs [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+
+        self.mqtt_publisher.send(firmware_update_req_msg).await?;
+        info!(
+            "Firmware update request is sent to child device. \
+            Details: child={child_id}, name={firmware_name}, version={firmware_version}, url={file_transfer_url}"
+        );
+
+        // A unique ID for timer. Once operation ID is possible to use, better to replace it.
+        let timer_id = digest(format!(
+            "{child_id}{firmware_name}{firmware_version}{firmware_url}"
```
Comment (albinsuresh) [inline plugins/c8y_firmware_plugin/src/download.rs:198]: Isn't the `firmware_url` unique enough for a firmware? Do we really need to append the `child_id`, `firmware_name` etc as well to make it "more unique"?

**Review by albinsuresh — COMMENTED** [review]
One aspect that I didn't fully understand from the impl is how the child devices would send their current firmware details to C8Y using the [set firmware message](https://cumulocity.com/guides/10.11.0/reference/smartrest-two/#115)? There was some contract for the `c8y-config-plugin` to send the initial config list to tedge/c8y. Something similar is needed here as well, right?

File: configuration/debian/c8y-firmware-plugin/prerm [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+# Stop the service only if systemctl is available
+if command -v systemctl >/dev/null; then
+    if systemctl is-active --quiet c8y-firmware-plugin.service; then
+         systemctl stop c8y-firmware-plugin.service
+    fi
+fi
+
+if [ -e /usr/bin/c8y_firmware_plugin ]; then
+    rm -f /usr/bin/c8y_firmware_plugin
+fi
```
[the remark this answers is not in the document]
Comment (reubenmiller) [reply configuration/debian/c8y-firmware-plugin/prerm:14]: Yes this section can be removed. It is only used in the other packages due to the package names changing (replacing underscores to dashes), hence we created a symlink in the `postinst` script, which then needs to be removed in the `prerm`.

tldr; Delete those three lines.

File: crates/common/tedge_config/src/tedge_config_cli/models/seconds.rs [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+fn conversion_from_longer_integer_fails() {
+    assert_matches!(
+        Seconds::try_from("18446744073709551616".to_string()),
+        Err(InvalidSecondsNumber { .. })
+    );
+}
+
+#[test]
+fn conversion_from_port_to_string() {
+    assert_matches!(TryInto::<String>::try_into(Seconds(1234)), Ok(seconds_str) if seconds_str == "1234");
```
[the remark this answers is not in the document]
Comment (rina23q) [reply crates/common/tedge_config/src/tedge_config_cli/models/seconds.rs:56]: Unfortunately your suggestion doesn't work. 

```
error: expected one of `)`, `,`, `...`, `..=`, `..`, or `|`, found `.`
  --> crates/common/tedge_config/src/tedge_config_cli/models/seconds.rs:58:18
   |
58 |         Ok("1234".to_string())
   |                  ^
   |                  |
   |                  expected one of `)`, `,`, `...`, `..=`, `..`, or `|`
   |                  help: missing `,`

```

File: plugins/c8y_firmware_plugin/src/child_device.rs (hunk 4 above)
Comment (rina23q) [reply plugins/c8y_firmware_plugin/src/child_device.rs:89]: This one is gone by later refactoring.

File: plugins/c8y_firmware_plugin/src/download.rs [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+use std::sync::Arc;
+use std::time::Duration;
+use tedge_api::OperationStatus;
+use tedge_utils::timers::Timers;
+use tokio::sync::Mutex;
+use tracing::error;
+use tracing::info;
+use tracing::warn;
+
+#[cfg(not(test))]
```
[the remark this answers is not in the document]
Comment (rina23q) [reply plugins/c8y_firmware_plugin/src/download.rs:37]: Unnecessary! :D

File: plugins/c8y_firmware_plugin/src/download.rs (hunk 8 above)
Comment (reubenmiller) [reply plugins/c8y_firmware_plugin/src/download.rs]: Having all 3 fragments (name, version, url) ensures that the timeout is only for this specific information. Just in case if people use the firmware name/version as some additional information which affects how the url is processed (or to use the name/version as re-processing logic, to enforce applying the same firmware binary again, but by just incrementing the version or changing the name).

File: plugins/c8y_firmware_plugin/src/firmware_manager.rs (hunk 1 above)
Comment (rina23q) [reply plugins/c8y_firmware_plugin/src/firmware_manager.rs]: We removed the abstraction of `FirmwareDownloadManager` together :)

File: plugins/c8y_firmware_plugin/src/download.rs (hunk 8 above)
Comment (rina23q) [reply plugins/c8y_firmware_plugin/src/download.rs]: Actually, this was already replaced by `{opration_id}`.

File: plugins/c8y_firmware_plugin/src/firmware_manager.rs [hunk 12]
```diff
@@ -0,0 +1,734 @@
...
+                if let Some(error_message) = &child_device_payload.reason {
+                    let failed_status_payload =
+                        DownloadFirmwareStatusMessage::status_failed(error_message.clone())?;
+                    mapped_responses.push(Message::new(&c8y_child_topic, failed_status_payload));
+                } else {
+                    let default_error_message =
+                        String::from("No fail reason provided by child device.");
+                    let failed_status_payload =
+                        DownloadFirmwareStatusMessage::status_failed(default_error_message)?;
+                    mapped_responses.push(Message::new(&c8y_child_topic, failed_status_payload));
```
Comment (albinsuresh) [inline plugins/c8y_firmware_plugin/src/firmware_manager.rs:400]:
```suggestion
                let failed_status_payload = DownloadFirmwareStatusMessage::status_failed(
                    child_device_payload
                        .reason
                        .unwrap_or_else(|| "No fail reason provided by child device.".into()),
                )?;
                mapped_responses.push(Message::new(&c8y_child_topic, failed_status_payload));
```

**Review by albinsuresh — COMMENTED** [review]
There's still scope for some more refactoring to simplify the error propagation. Failing operations in the cloud is done from multiple places which can be avoided with some error propagation refactoring.

File: crates/common/tedge_config/src/tedge_config_cli/settings.rs [hunk 13]
```diff
@@ -374,3 +374,16 @@ impl ConfigSetting for RunPathSetting {
 
     type Value = FilePath;
 }
+
+pub struct FirmwareTimeoutSetting;
+
+impl ConfigSetting for FirmwareTimeoutSetting {
+    const KEY: &'static str = "firmware.timeout";
```
[the remark this answers is not in the document]
Comment (rina23q) [reply crates/common/tedge_config/src/tedge_config_cli/settings.rs:381]: @reubenmiller Here I need your insight as it's UX. Which one do you want?

File: plugins/c8y_firmware_plugin/src/main.rs [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+    http_proxy.init().await?;
+    Ok(http_proxy)
+}
+
+#[tokio::main]
+async fn main() -> Result<(), anyhow::Error> {
+    let fw_plugin_opt = FirmwarePluginOpt::parse();
+
+    if fw_plugin_opt.init {
+        // Placeholder for the future enhancement
```
[the remark this answers is not in the document]
Comment (rina23q) [reply plugins/c8y_firmware_plugin/src/main.rs:65]: This was required by @reubenmiller.

File: plugins/c8y_firmware_plugin/src/firmware_manager.rs (hunk 12 above)
Comment (rina23q) [reply plugins/c8y_firmware_plugin/src/firmware_manager.rs:400]: I needed to change a bit as the suggested one cannot be complied.

File: plugins/c8y_firmware_plugin/src/main.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (reubenmiller) [reply plugins/c8y_firmware_plugin/src/main.rs]: Reason was to provide a consistent interface to all binaries...as most (all?) support the `--init` flag. Being consistent makes it easier for users to reason about

File: crates/common/tedge_config/src/tedge_config_cli/settings.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (reubenmiller) [reply crates/common/tedge_config/src/tedge_config_cli/settings.rs:381]: I prefer `firmware.timeout` as it is more concise, and `update` does not mean anything to a user, so it will just bring confusion (imho).

Though I can always be persuaded to change my mind if we are using similar style somewhere else (e.g. for consistency).

**Comment by reubenmiller** [comment]
> One aspect that I didn't fully understand from the impl is how the child devices would send their current firmware details to C8Y using the [set firmware message](https://cumulocity.com/guides/10.11.0/reference/smartrest-two/#115)? There was some contract for the `c8y-config-plugin` to send the initial config list to tedge/c8y. Something similar is needed here as well, right?

I think we can split the initial setting of the firmware when the child device starts in another PR, as we still need to think about the interface here. In the meantime the child device adapter/connector can update the firmware in the `c8y_Firmware` fragment via the `c8y/inventory/....` topic.

**Comment by reubenmiller** [comment]
@rina23q Now that the #1714 is merged, you will have rebase/merge the changes from `main` into your branch, as some of the integration test files are not present in your branch which causes the new `integration-test` workflow to fail.

File: plugins/c8y_firmware_plugin/src/main.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (albinsuresh) [reply plugins/c8y_firmware_plugin/src/main.rs]: Actually we found a use-case for this `--init` option to create all the necessary directory structures required by the plugin like the firmware cache and operation persistent store.

File: crates/common/tedge_config/src/tedge_config_cli/settings.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (albinsuresh) [reply crates/common/tedge_config/src/tedge_config_cli/settings.rs:381]: It definitely is concise, but at the same time, I felt that it's a bit too vague. In my head `firmware.timeout` sounds like the timeout for firmware, which doesn't make any sense. But something like `firmware.update.timeout` clearly conveys that it's the timeout for the firmware update operation. When we make the timeout for config management operations also configurable, I'd be adding `config.update.timeout` and `config.snapshot.timeout` which would be consistent with this firmware key.

TBH, I'd have gone for `child.config.update.timeout` to be completely explicit that it is applicable only for child devices, as in future we may want to provide such timeouts for these operations on the tedge device as well (say `tedge.config.update.timeout`). Being explicit now will help us avoid such conflicting/misleading keys in future. But, may be, I'm going too far into the future with this.

[the remark this answers is not in the document]
Comment (reubenmiller) [reply crates/common/tedge_config/src/tedge_config_cli/settings.rs:381]: You make valid points, we definitely need the `child` part the settings, as treating child devices will be different to the main device (when the main device is supported in the future).

My initial thinking for leaving out the `.update` part of the name was because there is only updates possible when firmware. But since other features have multiple actions (`update` and `snapshot`), then I will align with those as well.

So I will take @albinsuresh proposal `firmware.child.update.timeout`.

Example toml

```toml
[firmware]
child.update.timeout = 3600
```

File: plugins/c8y_firmware_plugin/src/firmware_manager.rs [hunk 15]
```diff
@@ -534,11 +521,45 @@ impl FirmwareManager {
         );
 
         if op_state == ActiveOperationState::Pending {
-            self.mqtt_client.published.send(executing_msg).await?;
+            self.mqtt_client
+                .published
+                .send(executing_msg)
+                .await
+                .map_err(MqttError::from)?;
```
Comment (albinsuresh) [inline plugins/c8y_firmware_plugin/src/firmware_manager.rs:528]: Why not just create `FirmwareManagementError` variant named `FromSendError(#[from] futures_channel::mpsc::SendError)` just like other errors, so that we can continue doing `.await?`

File: plugins/c8y_firmware_plugin/src/firmware_manager.rs [hunk 16]
```diff
@@ -534,11 +521,45 @@ impl FirmwareManager {
...
         }
 
-        self.mqtt_client.published.send(failed_msg).await?;
+        self.mqtt_client
+            .published
+            .send(failed_msg)
+            .await
+            .map_err(MqttError::from)?;
```
Comment (albinsuresh) [inline plugins/c8y_firmware_plugin/src/firmware_manager.rs:535]: Same comment as above.

File: plugins/c8y_firmware_plugin/src/firmware_manager.rs [hunk 17]
```diff
@@ -534,11 +521,45 @@ impl FirmwareManager {
...
+
+        Ok(())
+    }
+
+    // This function can be removed once we start using operation ID from c8y.
+    fn request_is_already_addressed(
```
Comment (albinsuresh) [inline plugins/c8y_firmware_plugin/src/firmware_manager.rs:541]:
```suggestion
    fn is_same_request_in_progress(
```
As we're just validating if the same request is already being processed using just the live operation files and not historical data. Even if the same request was processed earlier and completed, we won't be able to figure that out as we don't keep historical data.

File: plugins/c8y_firmware_plugin/src/firmware_manager.rs [hunk 18]
```diff
@@ -549,38 +570,26 @@ enum ActiveOperationState {
...
-            && (file_gid == tedge_gid || file_gid == root_gid)
-            && format!("{:o}", metadata.permissions().mode()).contains("644")
-        {
-            Ok(())
-        } else {
-            Err(FirmwareManagementError::InvalidFilePermission {
-                id: op_id.to_string(),
-            })
-        }
+    fn get_file_path(&self, op_id: &str) -> PathBuf {
```
Comment (albinsuresh) [inline plugins/c8y_firmware_plugin/src/firmware_manager.rs:591]:
```suggestion
    fn get_operation_file_path(&self, op_id: &str) -> PathBuf {
```
To be more contextual.

File: plugins/c8y_firmware_plugin/src/firmware_manager.rs [hunk 19]
```diff
@@ -598,16 +607,22 @@ pub struct FirmwareOperationEntry {
 }
 
 impl FirmwareOperationEntry {
-    fn create_file(&self) -> Result<(), FirmwareManagementError> {
-        let path = PersistentStore::get_file_path(&self.operation_id);
+    fn create_file(
```
Comment (albinsuresh) [inline plugins/c8y_firmware_plugin/src/firmware_manager.rs:610]: It feels more like this `create_file` or rather `create_operation_file` and `overwrite_file` functions should have been part of the `PresistentStore` as creating a file in the store should have been a function of that store.

File: plugins/c8y_firmware_plugin/src/main.rs [hunk 20]
```diff
@@ -21,7 +26,16 @@ use tedge_config::MqttPortSetting;
...
+use tedge_utils::file::create_directory_with_user_group;
 use tokio::sync::Mutex;
+use tracing::info;
+
+// TODO! We should make it configurable by tedge config later.
+const PERSISTENT_DIR_PATH: &str = "/var/tedge";
+
+pub const CACHE_DIR_NAME: &str = "cache";
+pub const FILE_TRANSFER_DIR_NAME: &str = "file-transfer";
+pub const PERSISTENT_STORE_DIR_NAME: &str = "firmware";
```
Comment (albinsuresh) [inline plugins/c8y_firmware_plugin/src/main.rs:38]: These constants better belong on the `firmware_manager` module itself so that they are portable when we migrate this plugin to the actor model with a different main.

File: plugins/c8y_firmware_plugin/src/main.rs [hunk 21]
```diff
@@ -21,7 +26,16 @@ use tedge_config::MqttPortSetting;
...
 use tedge_config::DEFAULT_TEDGE_CONFIG_PATH;
+use tedge_utils::file::create_directory_with_user_group;
 use tokio::sync::Mutex;
+use tracing::info;
+
+// TODO! We should make it configurable by tedge config later.
+const PERSISTENT_DIR_PATH: &str = "/var/tedge";
+
+pub const CACHE_DIR_NAME: &str = "cache";
+pub const FILE_TRANSFER_DIR_NAME: &str = "file-transfer";
```
Comment (albinsuresh) [inline plugins/c8y_firmware_plugin/src/main.rs:37]: TODO for later: This constant ideally should have been defined in some `tedge-api` crate so that all plugins using the file-transfer functionality could reuse it.

File: plugins/c8y_firmware_plugin/src/tests.rs [hunk 22]
```diff
@@ -0,0 +1,187 @@
...
+            "c8y/s/ds",
+            format!(
+                "515,{child_device_id},{firmware_name},{firmware_version},{cloud_firmware_url}"
+            )
+            .as_str(),
+        )
+        .await?;
+
+    // Since sha256 is generated by the downloaded file, can create RequestPayload after publishing 515.
+    // TODO! How to ignore the operation ID from the test result? It's randomly generated...
```
Comment (albinsuresh) [inline plugins/c8y_firmware_plugin/src/tests.rs:90]: We can do [partial JSON matching](https://docs.rs/assert-json-diff/latest/assert_json_diff/#partial-matching) which is supported by `assert_json_diff` used elsewhere in this project

File: plugins/c8y_firmware_plugin/src/firmware_manager.rs [hunk 23]
```diff
@@ -0,0 +1,734 @@
...
+#[cfg(test)]
+pub const FILE_TRANSFER_ROOT_PATH: &str = "/tmp/file-transfer";
+#[cfg(not(test))]
+pub const FILE_CACHE_DIR_PATH: &str = "/var/tedge/cache";
+#[cfg(test)]
+pub const FILE_CACHE_DIR_PATH: &str = "/tmp/cache";
+#[cfg(not(test))]
+pub const FIRMWARE_OPERATION_DIR_PATH: &str = "/var/tedge/firmware";
+#[cfg(test)]
+pub const FIRMWARE_OPERATION_DIR_PATH: &str = "/tmp/firmware";
```
[the remark this answers is not in the document]
Comment (rina23q) [reply plugins/c8y_firmware_plugin/src/firmware_manager.rs]: We fixed :)

**Comment by rina23q** [comment]
> @rina23q Now that #1714 is merged, you will have rebase/merge the changes from `main` into your branch, as some of the integration test files are not present in your branch which causes the new `integration-test` workflow to fail.

It's done.

File: crates/common/tedge_config/src/tedge_config_cli/settings.rs [hunk 24]
```diff
@@ -374,3 +374,16 @@ impl ConfigSetting for RunPathSetting {
 
     type Value = FilePath;
 }
+
+pub struct FirmwareTimeoutSetting;
```
[the remark this answers is not in the document]
Comment (rina23q) [reply crates/common/tedge_config/src/tedge_config_cli/settings.rs:378]: Changed to `FirmwareChildUpdateTimeoutSetting` as the key is `firmware.child.update.timeout`.

**Comment by rina23q** [comment]
I believe now this PR needs only minor fixes, therefore, change the state from Draft to Open.

File: plugins/c8y_firmware_plugin/src/tests.rs [hunk 25]
```diff
@@ -0,0 +1,775 @@
...
+
+const CHILD_DEVICE_ID: &str = "child-device";
+const FIRMWARE_NAME: &str = "fw-name";
+const FIRMWARE_VERSION: &str = "fw-version";
+
+#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
+#[serial_test::serial]
+async fn handle_request_child_device() -> anyhow::Result<()> {
+    let mut tmp_dir = TempTedgeDir::new();
+    create_required_directories(&mut tmp_dir);
```
Comment (albinsuresh) [inline plugins/c8y_firmware_plugin/src/tests.rs:24]: ~~Why not create these directories as part of the `start_firmware_manager` function itself?~~

File: plugins/c8y_firmware_plugin/src/tests.rs [hunk 26]
```diff
@@ -0,0 +1,775 @@
...
+            )
+            .as_str(),
+        )
+        .await?;
+
+    // Assert the c8y_Firmware operation status mapping to EXECUTING(501) and FAILED(502)
+    mqtt_tests::assert_received_all_expected(
+        &mut smartrest_messages,
+        TEST_TIMEOUT_MS,
+        &["501,c8y_Firmware", "502,c8y_Firmware"],
```
Comment (albinsuresh) [inline plugins/c8y_firmware_plugin/src/tests.rs:140]: It'd be nice if the failure reason(at least parts of it) in the 502 message could also be asserted to make sure that the code failed with the expected error and not with something else.

File: plugins/c8y_firmware_plugin/src/tests.rs [hunk 27]
```diff
@@ -0,0 +1,775 @@
...
```
Comment (albinsuresh) [inline plugins/c8y_firmware_plugin/src/tests.rs:201]: Assert failure reason if possible.

File: plugins/c8y_firmware_plugin/src/tests.rs [hunk 28]
```diff
@@ -0,0 +1,775 @@
...
```
Comment (albinsuresh) [inline plugins/c8y_firmware_plugin/src/tests.rs:323]: Assert failure reason.

File: plugins/c8y_firmware_plugin/src/tests.rs [hunk 29]
```diff
@@ -0,0 +1,775 @@
...
```
Comment (albinsuresh) [inline plugins/c8y_firmware_plugin/src/tests.rs:383]: Assert failure reason as timeout

File: plugins/c8y_firmware_plugin/src/tests.rs [hunk 30]
```diff
@@ -0,0 +1,775 @@
...
+            "c8y/s/ds",
+            format!(
+                "515,{CHILD_DEVICE_ID},{FIRMWARE_NAME},{FIRMWARE_VERSION},{cloud_firmware_url}"
+            )
+            .as_str(),
+        )
+        .await?;
+
+    // Assert the c8y_Firmware operation status mapping to EXECUTING(501) and FAILED(502)
+    mqtt_tests::assert_received_all_expected(
```
Comment (albinsuresh) [inline plugins/c8y_firmware_plugin/src/tests.rs:380]: Don't you need to make sure that the timeout period has elapsed before doing this assertion? With a `sleep` may be.

File: plugins/c8y_firmware_plugin/src/tests.rs [hunk 31]
```diff
@@ -0,0 +1,775 @@
...
+        .publish(
+            &format!("tedge/{CHILD_DEVICE_ID}/commands/res/firmware_update"),
+            &json!({
+                "status": "executing",
+                "id": operation_id,
+            })
+            .to_string(),
+        )
+        .await?;
+
```
Comment (albinsuresh) [inline plugins/c8y_firmware_plugin/src/tests.rs:559]: I'd assert the `501` message here to make sure that it is sent immediately in response to the `executing` response from the child.

File: plugins/c8y_firmware_plugin/src/tests.rs [hunk 32]
```diff
@@ -0,0 +1,775 @@
...
+        .await;
+
+    // Publish a successful RESPONSE from child device
+    broker
+        .publish(
+            &format!("tedge/{CHILD_DEVICE_ID}/commands/res/firmware_update"),
+            &json!({
+                "status": "successful",
+                "id": operation_id,
+                "reason": null
```
Comment (albinsuresh) [inline plugins/c8y_firmware_plugin/src/tests.rs:462]: When the reason is `null`, you can just skip that field completely.

File: plugins/c8y_firmware_plugin/src/tests.rs (hunk 25 above)
Comment (albinsuresh) [reply plugins/c8y_firmware_plugin/src/tests.rs:24]: Got the answer to that from tests like `handle_request_dir_cache_not_found` where you want to pick and choose the directories that are created.

Comment (rina23q) [reply plugins/c8y_firmware_plugin/src/tests.rs:24]: Exactly!

File: plugins/c8y_firmware_plugin/src/tests.rs (hunk 30 above)
Comment (rina23q) [reply plugins/c8y_firmware_plugin/src/tests.rs:380]: The firmware response timeout is 1 sec. We wait 5 seconds to check if all expected message are received inside assert. I don't think we need sleep.

File: plugins/c8y_firmware_plugin/src/tests.rs (hunk 32 above)
Comment (rina23q) [reply plugins/c8y_firmware_plugin/src/tests.rs:462]: Yes, also the payload is tested by unit tests.

File: crates/common/tedge_utils/src/file.rs [hunk 33]
```diff
@@ -58,6 +58,15 @@ pub fn create_directory_with_mode(dir: impl AsRef<Path>, mode: u32) -> Result<()
     perm_entry.create_directory(dir.as_ref())
 }
 
+pub fn create_file_with_mode(
```
[the remark this answers is not in the document]
Comment (didier-wenzek) [reply crates/common/tedge_utils/src/file.rs:67]: This can be done independently as this PR is already large.

That said I would suggest yet another direction (for later!). I agree with Albin that we cannot have a method per combination, but I also understand the need for something more explicit than a function with 3 optional arguments. So I would add `with_user`, `with_group` and `with_mode` method to `PermissionEntry`. So we could write: `PermissionEntry::new().with_mode(644).create_file(path, content)`.

File: plugins/c8y_firmware_plugin/src/firmware_manager.rs [hunk 34]
```diff
@@ -0,0 +1,754 @@
...
+            .await
+        {
+            self.fail_operation_in_cloud(&child_id, Some(&op_id), format!("{err}"))
+                .await?;
+        }
+
+        Ok(())
+    }
+
+    // TODO: Move some to functions
```
Comment (didier-wenzek) [inline plugins/c8y_firmware_plugin/src/firmware_manager.rs:236]: What do you mean?

**Review by didier-wenzek — COMMENTED** [review]
This sounds good and ready to be merged.

Just need to provide some proper help text.

File: plugins/c8y_firmware_plugin/src/firmware_manager.rs (hunk 34 above)
Comment (rina23q) [reply plugins/c8y_firmware_plugin/src/firmware_manager.rs]: I did this now with 813e745!

**Comment by rina23q** [comment]
> This can be done independently as this PR is already large.

> That said I would suggest yet another direction (for later!). I agree with Albin that we cannot have a method per combination, but I also understand the need for something more explicit than a function with 3 optional arguments. So I would add with_user, with_group and with_mode method to PermissionEntry. So we could write: PermissionEntry::new().with_mode(644).create_file(path, content).

It sounds good. I'd like to address in next PR!

**Review by albinsuresh — APPROVED** [review]
